nanoFramework.WebServer/WebServer.cs

```diff
@@ -665,6 +641,43 @@ protected virtual void InvokeRoute(CallbackRoutes route, HttpListenerContext con
             route.Callback.Invoke(null, new object[] { new WebServerEventArgs(context) });
         }
 
+        /// <summary>
+        /// Logs a message.
+        /// </summary>
+        /// <param name="message">The message to be logged.</param>
+        protected virtual void Log(string message)
```

Comment:
not sure it's optimal because at compilation Debug is removed when you're building in release. But the function will stay and the string transformation will stay as well. I'm not super sure the compiler  will remove everything. And as C# does not support "inline" function  in the sense of C++, I don't think this will really optimise the code.
See the MQTT lib, to avoid this issue because the log are formatted and requires a bit of love, I4ve added some #if DEBUG in the code whenever there is something to debug and format. So it's just totally skipped when not going thru.
Note for later: We should definitely test this behavior with the benchmark lib. Some function of this kind, empty inside, with string parameters and transformed.

Reply:
The idea was to be able to replace "Debug.Writeline" with own logger. For example, i'm using "serilog" like logger which is able to write logs into multiple sinks (like Console or SDcard).
But maybe better idea is to introduce events? The we can listen to those events?

Reply:
I like the logger and we have a standardize logger with API aligned with .NET. So we could use this one. Key element here as well is performance. I'm running some tests to see how the log function behave

Reply:
OK, just ran the tests:

```csharp
public void DoWithLog()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
        Log($"Some text and a is now: {a}");
    }
}

public void DoWithDebug()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
        Debug.WriteLine($"Some text and a is now: {a}");
    }
}

public void DoWitNothing()
{
    int a = 0;
    for (int i = 0; i < 1_000; i++)
    {
        a++;
    }
}

public void Log(string s)
{
    Debug.WriteLine(s);
}
```

Results I ran each of those 10 times), all in Release:

Do with Log: 00:01:28.2310360
Do with Debug : 00:00:00.0457140
Do with Nothing : 00:00:00.0455600

So as expected, there is a penalty for using a log function even if it's empty! Building in release is the equivalent of removing the Debug.WriteLine. As you can see in the test, even the empty function is called! So you have a penalty creating the string.

Bottom line:
* using the #if DEBUG is the way to go
* inside the block, you can use anything you want
* if you really want to use a logger, then the PN532 way is the way to go:
  *  check before calling the format function is the log level is higer than what you need to log.
 ``csharp
if (loglevel >= refLogLevel)
{
  Log($"Whatever you want to format {anything_here}");
]
 ```
This is going to improve performances!

Reply:
OK, for now rolling back to "Debug.Writeline"